### PR TITLE
Port to Python3 and other fixes [GCI]

### PR DESCRIPTION
**Pull request by srevinsaju**
The code has been heavily changed, The list_set union doesn't work on python3 due to change in override methods in classes. Tested on SugarDE. Linting done on most files

### CODE REVIEW

File: dobject.py [hunk 1]
```diff
@@ -507,114 +511,115 @@ class AddOnlySet:
     """
     def __init__(self, handler, initset = (), translator=empty_translator):
         self._logger = logging.getLogger('dobject.AddOnlySet')
-        self._set = set(initset)
-        
+        self._list = set(initset)
+        #self._list = list(initset)
```
Comment (quozl) [inline dobject.py]: Comment not needed.

File: dobject.py [hunk 2]
```diff
@@ -624,168 +629,170 @@ class AddOnlySortedSet:
...
     return value of get_history, and processing index information in add_history.
-    
+
     It is noteworthy that CausalHandler is in fact implemented on _top_ of
     UnorderedHandler.  The imposition of ordering does not require lower-level
     access to the network.  This fact of implementation may change in the
     future, but CausalObjects will not be able to tell the difference.
     """
-    _max64 = 2**64
+    max64 = 2**64
```
Comment (quozl) [inline dobject.py]: Is changed here but not where it is referenced.

File: dobject.py [hunk 3]
```diff
@@ -315,53 +317,53 @@ class HighScore:
...
         self._logger.debug("set_value_from_net " + str(val) + " " + str(score))
         if self._actually_set_value(val, score, tiebreaker):
             self._trigger()
-    
+
     def receive_message(self, message):
         self._logger.debug("receive_message " + str(message))
         if len(message) == 2: #Remote has break_ties=False
-            self._set_value_from_net(self._val_trans(message[0], False), self._score_trans(message[1], False), None)
+            self._list_value_from_net(self._val_trans(message[0], False), self._score_trans(message[1], False), None)
```
Comment (quozl) [inline dobject.py]: You change _set_value_from_net to _list_value_from_net here and a few lines later, but the function referenced is not changed.

Comment (srevinsaju) [reply dobject.py]: Fixed on https://github.com/sugarlabs/stopwatch/pull/7/commits/0b3a93d263c37e29be58b9390d63a0c924c4c840 ,

File: dobject.py (hunk 2 above)
Comment (srevinsaju) [reply dobject.py]: While fixing flake8, I removed an underscore on mistake, fixed on https://github.com/sugarlabs/stopwatch/pull/7/commits/0b3a93d263c37e29be58b9390d63a0c924c4c840

File: dobject.py (hunk 1 above)
Comment (srevinsaju) [reply dobject.py]: Removed comment

**Comment by srevinsaju** [comment]
@quozl , fixed it on https://github.com/sugarlabs/stopwatch/pull/7/commits/0b3a93d263c37e29be58b9390d63a0c924c4c840

**Comment by quozl** [comment]
Thanks.  Reviewed again.  Can you explain why self._set is changed to self._list in dobject?  There is so much about dobject that is based on the mathematical concept of sets, and you've changed a key variable name but I can't see why.   I've no trouble with self._set being a list datatype.  Why did it have to change?

My hesitation also stems from having to maintain a python2 branch into the future, and the amount of change in dobject.py is huge, for this reason and also whitespace.

**Comment by srevinsaju** [comment]
Yes @quozl , I have changed from `set` to `list` due to the following reasons
* The existing class `ListSet` uses a lot of deprecated method like __cmp__ and other lambda functions which is not compatible with Python3
* A new number of methods were added to the `ListSet` to give it the `settish` properties of having unique values, so on and so forth
* Changing from set to list, helps in keeping the source class as a base of `list` type than `set`. List with set properties gives the class `ListSet` .
` ListSet` = `builtin list` + `additional set functions`
Which is easier to maintain on python3 than 
`ListSet` = `builtin set` + `additional list functions`
Hope it is intuitive for you. It was the only workaround to the problem i faced. Sets and Lists can be natively methodized to use it under the sorted function. However not the ListSet class inherited from Javascript

PS: One more issue, If i continue using _set, it raises an error. 
`self._list is not found`
Because the dobject calls self._list.

**Comment by quozl** [comment]
Thanks.

@bemasc, this is a bit beyond my pay grade, and I'm worried I may accept regressions; have you time to review?

**Comment by srevinsaju** [comment]
@quozl, is it necessary to maintain a python2 branch? python 2 is at the verge of its death and new versions of GNU are not going to include python2 supposedly, idk. and the earlier made releases would be enough for the older sugar xo. it might be a part of your oversight plan, and I would like to know more on that :)

**Comment by quozl** [comment]
@srevinsaju, thanks for asking, yes, it is one of the things I'm paid to do by OLPC.  We made about three million laptops which have Sugar and Python 2.  We can't practically upgrade to Python 3 due to cost.  Cost of software engineering for firmware, kernel, drivers, configuration, and then operating system building.  For most of the activities I focus on, I do maintain _master_, _python2_ and also _fedora18_ branches, and try to port useful patches from _master_ into the other branches.  Minimising change helps with that.  For instance, your whitespace changes are good, and I'd take them, though it does mean I'll have to pick the changes out of your patch.  I'm fine with doing that.

File: dobject.py [hunk 4]
```diff
@@ -498,7 +498,7 @@ def _highscore_cb(self, val, score):
         for L in self._listeners:
             L(val)
 
-class AddOnlySet:
+class AddOnlySet(ListSet):
```
Comment (bemasc) [inline dobject.py:501]: Was there a reason for this change to inherit from ListSet?  It seems like a good idea, but this probably makes all of the copying of operator methods (like `self.__and__`) unnecessary.

**Review by bemasc — COMMENTED** [review]
It's been many years since I looked at this, so I don't have any special insight, but thanks for maintaining it.   The GObject and _thread changes look a little scary but probably fine.

**Comment by quozl** [comment]
@bemasc, thanks.

@srevinsaju, I know you don't have resources to test collaboration, but I think testing collaboration is important, because that's what the GroupThink component is for.  We will hvae to wait for someone to do it.

**Pull request merged**